rust/src/serialization.rs

```diff
@@ -626,11 +664,222 @@ impl DeserializeEmbeddedGroup for TransactionOutput {
+                }
+            }
+            let datum = match raw.unsigned_integer()? {
+                0 => DataOption::DataHash(DataHash::deserialize(raw)?),
+                1 => {
+                    match raw.tag()? {
+                        //bytes string tag
+                        24 => {
+                            let data = (|| -> Result<_, DeserializeError> {
+                                Ok(from_bytes(&raw.bytes()?)?)
+                            })().map_err(|e| e.annotate("PlutusData"))?;
+                            DataOption::Data(data)
+                        },
+                        tag => {
+                            return Err(DeserializeFailure::TagMismatch {
+                                found: tag,
+                                expected: 24,
+                            }.into());
+                        }
+                    }
+                }
+                n => return Err(DeserializeFailure::FixedValueMismatch {
+                    found: Key::Uint(n),
+                    expected: Key::Uint(0),
+                }.into()),
+            };
+            if let cbor_event::Len::Indefinite = len {
+                if raw.special()? != CBORSpecial::Break {
+                    return Err(DeserializeFailure::EndingBreakMissing.into());
+                }
+            }
+            Ok(datum)
+        })().map_err(|e| e.annotate("DataOption"))
+    }
+}
+
+impl cbor_event::se::Serialize for DataOption {
+    fn serialize<'a, W: Write + Sized>(&self, serializer: &'a mut Serializer<W>) -> cbor_event::Result<&'a mut Serializer<W>> {
+        serializer.write_array(cbor_event::Len::Len(2))?;
+        match &self {
+            DataOption::DataHash(data_hash) => {
+                serializer.write_unsigned_integer(0)?;
+                data_hash.serialize(serializer)?;
+            },
+            DataOption::Data(data) => {
+                serializer.write_unsigned_integer(1)?;
+                let bytes = data.to_bytes();
+                serializer.write_tag(24)?
+                    .write_bytes(&bytes)?;
+            }
+        }
+        Ok(serializer)
+    }
+}
+
+impl Deserialize for ScriptRefEnum {
+    fn deserialize<R: BufRead + Seek>(raw: &mut Deserializer<R>) -> Result<Self, DeserializeError> {
+        (|| -> Result<_, DeserializeError> {
+            let len = raw.array()?;
+            if let cbor_event::Len::Len(n) = len {
+                if n != 2 {
+                    return Err(DeserializeFailure::CBOR(cbor_event::Error::WrongLen(2, len, "[id, native_or_putus_script]")).into())
+                }
+            }
+            let script_ref = match raw.unsigned_integer()? {
+                0 => ScriptRefEnum::NativeScript(NativeScript::deserialize(raw)?),
+                1 => ScriptRefEnum::PlutusScriptV1(PlutusScript::deserialize(raw)?),
+                //TODO: add plutus v2 tag
```

Comment:
(just note to future us) // todo, to be updated to check the version tag within the `PlutusScript` itself. The usage of `PlutusScriptV2` enum must be also removed here